tests/kernel/sched/schedule_api/src/test_sched_pending.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <ksched.h>
+#include <spinlock.h>
+#include <wait_q.h>
+#include <ztest.h>
+
+struct sched_pending { 
+	_wait_q_t waitq;
+	struct k_spinlock lock;
+}pending;
+
+#define STACK_SIZE	(1024 + CONFIG_TEST_EXTRA_STACKSIZE)
+
+K_THREAD_STACK_DEFINE(test_stack, STACK_SIZE);
+
+static struct k_thread tdata;
+static int num;
+static k_tid_t tid;
+
+/* action for self-defined assert handler. */
+static ZTEST_DMEM volatile bool valid_assert;
+
+static inline void set_assert_valid(bool valid)
+{
+	valid_assert = valid;
+}
+
+static void action_after_assert_fail(void)
+{
+	k_thread_abort(tid);
+	
+	ztest_test_pass();
+}
+
+#ifdef CONFIG_ASSERT_NO_FILE_INFO
+void assert_post_action(void)
+#else
+void assert_post_action(const char *file, unsigned int line)
+#endif
+{
+#ifndef CONFIG_ASSERT_NO_FILE_INFO
+	ARG_UNUSED(file);
+	ARG_UNUSED(line);
+#endif
+
+	printk("Caught an assert.\n");
+
+	if (valid_assert) {
+		valid_assert = false; /* reset back to normal */
+		printk("Assert error expected as part of test case.\n");
+
+		/* do some action after fatal error happened */
+		action_after_assert_fail();
+	} else {
+		printk("Assert failed was unexpected, aborting...\n");
+#ifdef CONFIG_USERSPACE
+	/* User threads aren't allowed to induce kernel panics; generate
+	 * an oops instead.
+	 */
+		if (_is_user_context()) {
+			k_oops();
+		}
+#endif
+		k_panic();
+	}
+}
+
+static void thread_handler1(void *p1, void *p2, void *p3)
+{
+	int times = (int)p1;
+	
+	z_waitq_init(&pending.waitq);
+	for(int i = 0; i < times; i++){
+		k_spinlock_key_t key = k_spin_lock(&pending.lock);
+		z_pend_thread(_current, &pending.waitq, K_FOREVER);
+		z_reschedule(&pending.lock, key);
+	}
+}
+
+static void thread_handler2(void *p1, void *p2, void *p3)
+{
+	z_waitq_init(&pending.waitq);
+
+	unsigned int key = irq_lock();
+	z_pend_curr_irqlock(key, &pending.waitq, K_FOREVER);
+}
+
+static void thread_handler3(void *p1, void *p2, void *p3)
+{
+	num++;
```

Comment:
I'm wondering if we should add a kind of lock here?

Reply:
This is really a problem that needs attention. I have changed code and used atomic variable here.